url0d38a10ee7_diegopliebana_GEBT_tgz-pJ8-ch_idsia_scenarios_test_EvolveWithChangingSeedsJ8/src/grammaticalbehaviors/GEBT_Mario/Map.java

```diff
@@ -0,0 +1,1047 @@
+
+    }
+
+    public void flushEnemiesAndItems()
+    {
+        m_enemyCollection.clear();
+        m_itemCollection.clear();
+    }
+
+    //X is defined by m_currentXWritting
+    public void writeEnemy(int a_y, byte a_data)
+    {
+        //Check boundaries
+        if(m_currentXWritting == -1 || a_y < 0 || a_y >= LEVEL_HEIGHT)
+            return;
+
+        if(a_data == Sprite.KIND_MUSHROOM || a_data == Sprite.KIND_FIRE_FLOWER)
+        {
+            if(!m_itemCollection.containsKey(m_currentXWritting))
+            {
+                m_itemCollection.put(m_currentXWritting, new Vector<Item>());
+            }
+            m_itemCollection.get(m_currentXWritting).add(new Item(m_currentXWritting, a_y, a_data));
+        }
+        else
+        {
+            if(!m_enemyCollection.containsKey(m_currentXWritting))
+            {
+                m_enemyCollection.put(m_currentXWritting, new Vector<Enemy>());
+            }
+            m_enemyCollection.get(m_currentXWritting).add(new Enemy(m_currentXWritting, a_y, a_data));
+        }
+    }
+
+    public boolean checkForEnemyType(int a_xSrc, int a_xDest, int a_ySrc, int a_yDest, int a_type)
+    {
+        int min_x = Math.min(a_xSrc, a_xDest);
+        int min_y = Math.min(a_ySrc, a_yDest);
+        int max_x = Math.max(a_xSrc, a_xDest);
+        int max_y = Math.max(a_ySrc, a_yDest);
+
+        if(min_x < 0) min_x = 0;
+        if(max_y >= 15) max_y = 14;
+        if(min_y < 0) min_y = 0;
+
+
+        for(int x = min_x; x <= max_x; ++x)
+        {
+            for(int y = min_y; y <= max_y; ++y)
+            {
+                if(m_enemyCollection.containsKey(x))
+                {
+                    Vector<Enemy> enemies = m_enemyCollection.get(x);
+                    for(int i = 0; i < enemies.size(); ++i)
+                    {
+                        Enemy en = enemies.elementAt(i);
+                        if(en.m_type == a_type && en.m_y == y)
+                        {
+                            return true;
+                        }
+                    }
+                }
+            }
+        }
+        return false;
+        
+    }
+
+
+    public void updateLastCell()
+    {
+        //Update last cell written
+        if(m_currentXWritting > m_lastCell)
+            m_lastCell = m_currentXWritting;
+    }
+
+    public byte read(int a_x, int a_y)
+    {
+        //Check boundaries
+        if(a_y < 0 || a_y >= LEVEL_HEIGHT)
+            return -1;
+
+        //Get the block for this position
+        int blockIndex = (int)(a_x / MAP_BLOCK_LENGTH);
+        int xInBlock = (int)(a_x % MAP_BLOCK_LENGTH);
+        if(blockIndex < m_levelMap.size())
+        {
+            return getBlock(blockIndex).read(xInBlock,a_y);
+        }
+        
+        return -1;
+    }
+
+    public void dumpToFile(String filename)
+    {
+        try
+        {
+            FileWriter fw = new FileWriter(filename);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *fw*. The resource is referenced at lines 284, 293, 297. The resource is closed at line 301. There are other execution paths that do not contain closure statements, for example, when MapBlock.dumpToFile throws an exception. Ensure *fw* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).